torch/csrc/api/include/torch/data/htk/htk_parser.hpp

```diff
@@ -0,0 +1,175 @@
+
+  template <typename T>
+  void parse_data(const std::string &file_name, const std::string &target_name, const size_t chunk_size, std::vector<Utterance> &data)
+  {
+    assert(!m_chunk_file_stream.is_open());
+    m_chunk_file_stream.open(file_name.data(), ios::in | ios::binary);
+
+    // Read header string
+    ReadString(m_chunk_file_stream, target_name);
+
+    // Read version number
+    std::optional<int32_t> version = ReadDebugData<int32_t>(m_chunk_file_stream);
+
+    int element_count = 0;
+
+    for (int element_count = 0; element_count < chunk_size; ++element_count)
+    {
+
+      std::optional<int32_t> id = ReadDebugData<int32_t>(m_chunk_file_stream);
+
+#ifdef _DEBUG
+      if (id && id.value() != element_count)
+      {
+        std::string error_msg = "File " + file_name + "is corrupted. The number " + std::to_string(element_count) + " element is missing";
+        throw std::runtime_error(error_msg);
+      }
+#endif
+      int32_t data_byte_size = ReadData<int32_t>(m_chunk_file_stream);
+
+      if (data_byte_size % sizeof(T) != 0)
+      {
+        std::string error_msg = "data size in byte is not a multiplication of " + std::string(typeid(T).name());
+        throw std::runtime_error(error_msg);
+      }
+
+      int32_t data_size = data_byte_size / sizeof(T);
+      void *data_ptr;
+
+      if (target_name == "Feature")
+      {
+        data[element_count].feature.resize(data_size);
+        data_ptr = reinterpret_cast<void *>(data[element_count].feature.data());
+      }
+      else if (target_name == "Lattice")
+      {
+        data[element_count].lattice.resize(data_size);
+        data_ptr = reinterpret_cast<void *>(data[element_count].lattice.data());
+      }
+      else if (target_name == "Label")
+      {
+        data[element_count].label.resize(data_size);
+        data_ptr = reinterpret_cast<void *>(data[element_count].label.data());
+      }
+      else
+      {
+        std::string error_msg = "unknown field " + target_name;
+        throw std::runtime_error(error_msg);
+      }
+      ReadDataArray(m_chunk_file_stream, reinterpret_cast<T *>(data_ptr), data_byte_size);
+    }
+    m_chunk_file_stream.close();
+  }
+
+  template <typename T>
+  static T ReadData(ifstream &reader)
+  {
+    T result;
+    reader.read(reinterpret_cast<char *>(&result), sizeof(result));
+
+    return is_little_endian_ ? swap_endian<T>(result) : result;
+  }
+
+  template <typename T>
+  static std::optional<T> ReadDebugData(ifstream &reader)
+  {
+#ifdef _DEBUG
+    T result;
+    reader.read(reinterpret_cast<char *>(&result), sizeof(result));
+    return is_little_endian_ ? swap_endian<T>(result) : result;
+#else
+    reader.ignore(sizeof(T));
+    return {};
+#endif
+  }
+
+  template <typename T>
+  static void ReadDataArray(ifstream &reader, T *value, size_t byte_length)
+  {
+    size_t value_size = sizeof(T);
+    while (byte_length > 0)
+    {
+      T v;
+      reader.read(reinterpret_cast<char *>(&v), value_size);
```

Comment:
What could we do to improve perf of this. Could we avoid the endin_ness check for each element. Another is can we load the entire array at once instead of per element?

Reply:
>Could we avoid the endin_ness check for each element
Sadly no. I was trying to see if the stream can be smarter on the byte reading order, but it turns out it can only do sequential loading.
>Another is can we load the entire array at once instead of per element?
Yes, for this we can do :)